README.md
# pageclip-ruby

A simple ruby gem to interface with the Pageclip API.

## Example

```
require 'pageclip'

Pageclip.configure do |config|
  config.api_key = 'your-secret-api-key'
end

begin
  data = Pageclip.screenshot('http://www.example.com')
  File.open('screenshot.png') do |f|
    f.write(data)
  end
  puts 'The screenshot is done!'
rescue Pageclip::Timeout
  puts 'Sorry I couldn't find a screenshot
rescue Pageclip::UnauthorizedError
  puts 'You don't have a valid API key'
rescue Pageclip::RateLimitedError
  puts 'Too many requests, wait before trying again'
rescue Pageclip::ServiceUnavailableError
  puts 'Somethings wrong, wait before trying again'
rescue Pageclip::ScreenshotError
  puts "Somethings wrong with the site you're requesting a screenshot for"
rescue Pageclip::Error
  puts 'Unknown error'
end
```

## Global configuration

 * api_key - Your API key.
 * job_defaults - This set of options will be used for all requests and
   can be overriden by specifying any of the properties as options to
   #screenshot.

## Job Parameters

 * viewport_width, viewport_height - Controls the size of viewport of
   the browser in pixels. It defaults to 1024x768 (permitted: 0-1024).
 * canvas_width, canvas_height - Controls the size in pixels of the
   entire page. If undefined it will screenshot the entire page.
   (permitted: 0-10000)
 * thumbnail_width, thumbnail_height - Controls the size in pixels of the
   final screenshot. If undefined it will result in a screenshot at the
   same size as the canvas. (permitted: 0-3000)
 * secret - Sets a custom header ('X-Secret') so you can authenticate
   the request coming to one of your private pages. If undefined the
   header will not be sent. (any string less than 256 characters)
 * timeout - Wait x number of seconds from starting to load the page to
   take a screenshot. If undefined it will load as soon as the page as
   finished loading (permitted: 0-55).


CHANGELOG.md
0.0.1 / 2013-02-21
==================

  * Initial version


spec/spec_helper.rb
require 'rubygems'
require 'bundler/setup'
require 'pageclip'

require 'webmock/rspec'

RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
  config.filter_run :focus

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = 'random'
end


spec/pageclip_spec.rb
require 'spec_helper'

describe 'Pageclip' do
  subject do
    Pageclip
  end

  let(:api_key) { 'key' }
  let(:url) { 'http://www.example.com' }
  let(:secret) { 'a' }

  describe 'configuring' do
    it 'can set the api key' do
      subject.configuration.reset!
      subject.configure do |config|
        config.api_key = api_key
        config.job_defaults = {}
      end
      subject.configuration.api_key.should eq(api_key)
    end
  end

  describe 'taking a screenshot' do
    context 'with configuration' do
      before do
        subject.configuration.reset!
        subject.configure do |config|
          config.api_key = api_key
        end
      end

      it 'can request with just a url' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:headers => {'User-Agent' => "Pageclip Ruby #{Pageclip::VERSION}"}).
          with(:query => {'url' => url, 'api_key' => api_key}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          with(:headers => {'User-Agent' => "Pageclip Ruby #{Pageclip::VERSION}"}).
          to_return(:status => 301, :headers => { :location => 'https://s3.amazonaws.com/bucket/1.png' })

        image_url = subject.screenshot(url)
        screenshot.should have_been_requested
        result.should have_been_requested
        image_url.should eq('https://s3.amazonaws.com/bucket/1.png')
      end

      it 'can request with a url and options' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key, 'secret' => secret}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          to_return(:status => 301, :headers => { :location => 'https://s3.amazonaws.com/bucket/1.png' })

        image_url = subject.screenshot(url, :secret => secret)
        screenshot.should have_been_requested
        screenshot.should have_been_requested
        result.should have_been_requested
        image_url.should eq('https://s3.amazonaws.com/bucket/1.png')
      end

      it 'handles timeouts' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).to_raise(Timeout::Error)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::TimeoutError)
        screenshot.should have_been_requested
      end
      it 'handles timeouts' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).to_raise(Timeout::Error)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::TimeoutError)
        screenshot.should have_been_requested
      end

      it 'handles unauthorized errors' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).to_return(:status => 403)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::UnauthorizedError)
        screenshot.should have_been_requested
      end

      it 'handles rate limit errors' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).to_return(:status => 429)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::RateLimitedError)
        screenshot.should have_been_requested
      end

      it 'handles service unavailable errors' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).to_return(:status => 503)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::ServiceUnavailableError)
        screenshot.should have_been_requested
      end

      it 'handles service unavailable errors on results' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          to_return(:status => 503)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::ServiceUnavailableError)
        screenshot.should have_been_requested
      end

      it 'handles screenshot errors' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').to_return(:status => 410)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::ScreenshotError)
        screenshot.should have_been_requested
        result.should have_been_requested
      end

      it 'handles fewer than 3 EOF errors' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          to_raise(EOFError).then.to_return(:status => 301, :headers => { :location => 'http://s3.amazonaws.com/bucket/1.png' })
        Kernel.should_receive(:sleep).with(1)

        subject.screenshot(url).should eq('http://s3.amazonaws.com/bucket/1.png')
        screenshot.should have_been_requested
        result.should have_been_requested.times(2)
      end

      it 'handles errors on the third EOF error' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          to_raise(EOFError).then.to_raise(EOFError).then.to_raise(EOFError)
        Kernel.should_receive(:sleep).with(1)
        Kernel.should_receive(:sleep).with(2)
        Kernel.should_receive(:sleep).with(3)

        expect { subject.screenshot(url) }.to raise_error(Pageclip::Error)
        screenshot.should have_been_requested
        result.should have_been_requested.times(4)
      end
    end

    context 'with default job options' do
      before do
        subject.configuration.reset!
        subject.configure do |config|
          config.api_key = api_key
          config.job_defaults = {
            :secret => secret
          }
        end
      end

      it 'can request with just a url' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key, 'secret' => secret}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          to_return(:status => 301, :headers => { :location => 'https://s3.amazonaws.com/bucket/1.png' })

        image_url = subject.screenshot(url)
        screenshot.should have_been_requested
        result.should have_been_requested
        image_url.should eq('https://s3.amazonaws.com/bucket/1.png')
      end

      it 'can override a property' do
        screenshot = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key, 'secret' => 'b'}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        result = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          to_return(:status => 301, :headers => { :location => 'https://s3.amazonaws.com/bucket/1.png' })

        image_url = subject.screenshot(url, :secret => 'b')
        screenshot.should have_been_requested
        result.should have_been_requested
        image_url.should eq('https://s3.amazonaws.com/bucket/1.png')
      end
    end
    context 'with a logger' do
      before do
        subject.configuration.reset!
        subject.configure do |config|
          config.api_key = api_key
          config.logger = logger
        end
      end

      let(:logger) {
        class MyLogger
          attr_reader :messages

          def initialize
            @messages = []
          end

          def info(message)
            @messages << message
          end
        end
        MyLogger.new
      }

      it 'captures a message on success' do
        stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key, 'secret' => 'b'}).
          to_return(:status => 302, :headers => { :location => 'http://api.pageclip.io/v1/screenshots/1' })
        stub_request(:get, 'http://api.pageclip.io/v1/screenshots/1').
          to_return(:status => 301, :headers => { :location => 'https://s3.amazonaws.com/bucket/1.png' })

        subject.screenshot(url, :secret => 'b')
        logger.messages.length.should eq(1)
        logger.messages[0].should match(/\[Pageclip 301 [0-9\.]+s\] Requested #{url}/)
      end

      it 'captures a message on failed response' do
        stub = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key, 'secret' => 'b'}).to_return(:status => 403)

        expect { subject.screenshot(url, :secret => 'b') }.to raise_error(Pageclip::UnauthorizedError)
        logger.messages.length.should eq(1)
        logger.messages[0].should match(/\[Pageclip 403 [0-9\.]+s\] Requested #{url}/)
      end

      it 'captures a message on exception' do
        stub = stub_request(:get, 'http://api.pageclip.io/v1/screenshots/').
          with(:query => {'url' => url, 'api_key' => api_key, 'secret' => 'b'}).to_raise(Timeout::Error)

        expect { subject.screenshot(url, :secret => 'b') }.to raise_error(Pageclip::TimeoutError)
        logger.messages.length.should eq(1)
        logger.messages[0].should match(/\[Pageclip - \?+s\] Requested #{url}/)
      end
    end
  end
end


spec/pageclip/configuration_spec.rb
require 'logger'
require 'spec_helper'

describe 'Pageclip::Configuration' do
  subject do
    Pageclip::Configuration.new
  end

  it 'can set the api key' do
    subject.api_key = 'asdf'
    subject.api_key.should eq('asdf')
  end

  it 'can set the job defaults' do
    subject.job_defaults = { :timeout => 30 }
    subject.job_defaults[:timeout].should eq(30)
  end

  it 'can set the logger' do
    subject.logger = Logger.new(STDOUT)
    subject.logger.should respond_to(:info)
  end

  it 'has a default api endpoint' do
    subject.api_endpoint.should eq('http://api.pageclip.io')
  end

  it 'has a default client timeout' do
    subject.client_timeout.should eq(61)
  end
end


lib/pageclip.rb
require 'net/http'
require 'uri'
require 'benchmark'

require 'pageclip/version'
require 'pageclip/configuration'
require 'pageclip/errors'

module Pageclip
  class << self
    def configure
      yield(configuration)
    end

    # Public: The current configuration for the Pageclip service.
    #
    # Returns a Pageclip::Configuration representing the current
    # configuration.
    def configuration
      @configuration ||= Pageclip::Configuration.new
    end

    # Public: Requests a screenshot from synchronously.
    #
    # options:
    #   :canvas_width
    #   :canvas_height
    #   :viewport_width
    #   :viewport_height
    #   :thumbnail_width
    #   :thumbnail_height
    #   :timeout
    #   :secret
    #
    def screenshot(url, options={})
      time = nil
      response = nil

      begin
        Timeout::timeout(@configuration.client_timeout) {
          request_options = @configuration.job_defaults || {}
          request_options[:api_key] = @configuration.api_key
          request_options.merge!(options)
          request_options[:url] = url

          time = Benchmark.realtime do
            response = get("#{@configuration.api_endpoint}/v1/screenshots/", request_options)
          end

          if response.code == "403"
            raise Pageclip::UnauthorizedError
          elsif response.code == "429"
            raise Pageclip::RateLimitedError
          elsif response.code == "503"
            raise Pageclip::ServiceUnavailableError
          elsif response.code == "302"
            time += Benchmark.realtime do
              begin
                response = get(response['location'])
              rescue EOFError
                if attempt = ( attempt || 1) and attempt <= 3
                  Kernel.sleep(attempt)
                  attempt += 1
                  retry
                end
              end
            end
          end

          if response.code == "410" || response.code == "202"
            raise Pageclip::ScreenshotError
          elsif response.code == "503"
            raise Pageclip::ServiceUnavailableError
          elsif response.code == "301"
            response['location']
          else
            raise Pageclip::Error
          end
        }
      rescue Timeout::Error
        raise Pageclip::TimeoutError
      ensure
        log("[Pageclip #{response ? response.code : '-'} #{time ? time : '?'}s] Requested #{url}")
      end
    end

  private
    def log(message)
      return unless @configuration.logger

      @configuration.logger.info(message)
    end

    def get(url, query={})
      uri = URI.parse(url)
      unless query.empty?
        uri.query = query.map { |k,v| "#{URI.escape(k.to_s)}=#{URI.escape(v.to_s)}" }.join("&")
      end

      http = Net::HTTP.new(uri.host, uri.port)
      request = Net::HTTP::Get.new(uri.request_uri)
      request['User-Agent'] = "Pageclip Ruby #{VERSION}"
      http.request(request)
    end
  end
end
